tachyon/crypto/hashes/sponge/poseidon/poseidon_config.h

```diff
@@ -173,6 +183,49 @@ struct PoseidonConfig : public PoseidonConfigBase<F> {
       optimized_ark.row(i) = mds_inverse * ark.row(i).transpose();
     }
   }
+
+  constexpr static void ComputeSparseMatrices(
+      size_t partial_rounds, const math::Matrix<F>& mds,
+      math::Matrix<F>& pre_sparse_mds,
+      std::vector<SparseMDSMatrix<F>>& sparse_matrices) {
+    math::Matrix<F> mds_transpose = mds.transpose();
+    math::Matrix<F> acc = mds_transpose;
+    sparse_matrices =
+        base::CreateVector(partial_rounds, [&mds_transpose, &acc]() {
+          math::Matrix<F> m_prime;
+          SparseMDSMatrix<F> m_prime_prime;
+          Factorize(acc, m_prime, m_prime_prime);
+          acc = mds_transpose * m_prime;
+          return m_prime_prime;
+        });
+    std::reverse(sparse_matrices.begin(), sparse_matrices.end());
+    pre_sparse_mds = acc.transpose();
+  }
+
+  // See Section B in Supplementary Material
+  // https://eprint.iacr.org/2019/458.pdf Factorizes an MDS matrix |M| into
+  // |M'| and |M''| where |M = M' *  M''|. Resulted |M''| matrices are the
+  // sparse ones while |M'| will contribute to the accumulator of the process.
+  constexpr static void Factorize(const math::Matrix<F>& mds,
+                                  math::Matrix<F>& m_prime,
+                                  SparseMDSMatrix<F>& m_prime_prime) {
+    // Given |(t - 1 * t - 1)| MDS matrix called |m_hat| constructs the matrix
+    // in form |[[1 | 0], [0 | m_hat]]|.
+    Eigen::Block<const math::Matrix<F>> m_hat =
+        mds.block(1, 1, mds.rows() - 1, mds.cols() - 1);
+    m_prime = math::Matrix<F>::Identity(mds.rows(), mds.cols());
+    m_prime.block(1, 1, mds.rows() - 1, mds.cols() - 1) = m_hat;
+
+    // Given |(t - 1)| sized |w_hat| vector constructs the matrix in form
```

Comment:
to match with the comment above
```suggestion
  // |(t - 1)| sized vector called |w_hat|  is denoted as |[[m₀,₀ | m₀,ᵢ], [w_hat | identity]]|.